sdks/java/core/src/main/java/org/apache/beam/sdk/values/BeamRecord.java

```diff
@@ -237,27 +200,27 @@ public void setDataValues(List<Object> dataValues) {
     this.dataValues = dataValues;
   }
 
-  public BeamSqlRowType getDataType() {
+  public BeamRecordType getDataType() {
     return dataType;
   }
 
-  public void setDataType(BeamSqlRowType dataType) {
+  public void setDataType(BeamRecordType dataType) {
     this.dataType = dataType;
   }
 
-  public void setNullFields(List<Integer> nullFields) {
-    this.nullFields = nullFields;
+  public BitSet getNullFields() {
+    return nullFields;
   }
 
-  public List<Integer> getNullFields() {
-    return nullFields;
+  public void setNullFields(BitSet nullFields) {
```

Comment:
Yes, we must store it in the encoded representation. But between the choice of setting it to all-true in the constructor and updating it as fields are mutated (we're not doing this correctly btw, see setDataValues and possibly elsewhere) vs. computing it on decode, I prefer the locality of the latter as it's just as cheap (and really cheap compared to the actual serialization), but your call.